### PR TITLE
SH-683: Setting Redis cache while readingGroup

### PR DESCRIPTION
**Pull request by aishwa8141**
Used two Identifier

1.  groupId: group and activity information

2.  groupId_members: group members info

### CODE REVIEW

File: group-actors/src/main/java/org/sunbird/actors/CacheActor.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        getCache(request);
+        break;
+      case "delCache":
+        delCache(request);
+      default:
+        onReceiveUnsupportedMessage("CacheActor");
+    }
+  }
+
+  private void setCache(Request request) {
```
Comment (indrajra) [inline group-actors/src/main/java/org/sunbird/actors/CacheActor.java:42]: All these methods have a log line what id they're trying with.

File: group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java [hunk 2]
```diff
@@ -35,15 +39,36 @@ public void onReceive(Request request) throws Throwable {
...
-  private void readGroup(Request actorMessage) throws BaseException {
+  private void readGroup(Request actorMessage) throws Exception {
     logger.info("ReadGroup method call");
     String groupId = (String) actorMessage.getRequest().get(JsonKey.GROUP_ID);
     List<String> requestFields = (List<String>) actorMessage.getRequest().get(JsonKey.FIELDS);
-
+    GroupResponse groupResponse;
+    String groupInfo = cacheUtil.getCache(groupId);
+    if (StringUtils.isNotEmpty(groupInfo)) {
+      groupResponse = JsonUtils.deserialize(groupInfo, GroupResponse.class);
```
Comment (indrajra) [inline group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java:49]: I see deserialization here and then in 70 to convert to a Map. Can we do one thing only - at least for Redis cache reads?

File: group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java [hunk 3]
```diff
@@ -35,15 +39,36 @@ public void onReceive(Request request) throws Throwable {
...
+    } else {
+      groupResponse = groupService.readGroup(groupId);
+      cacheUtil.setCache(groupId, JsonUtils.serialize(groupResponse), self());
+    }
+    if (requestFields.contains(JsonKey.MEMBERS)) {
+      String groupMember = cacheUtil.getCache(groupId + "_" + JsonKey.MEMBERS);
```
Comment (indrajra) [inline group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java:55]: This concatenation of groupId _ members literal must itself be encapsulated in a function. Example: constructRedisId(groupId)

File: service/conf/application.conf [hunk 4]
```diff
@@ -345,6 +345,11 @@ db {
 }
 play.filters.disabled += play.filters.csrf.CSRFFilter
 
+# Redis Configuration
+redis.host=localhost
```
Comment (indrajra) [inline service/conf/application.conf:349]: System env var

File: service/conf/application.conf [hunk 5]
```diff
@@ -480,6 +485,11 @@ thisActorSystem {
             nr-of-instances = 5
             dispatcher = group-dispatcher
           }
+        /CacheActor{
+          router = smallest-mailbox-pool
+          nr-of-instances = 5
```
Comment (indrajra) [inline service/conf/application.conf:490]: Group-dispatcher actor instances need some higher numbers.
read - 15
cache - 15

create - 10
search - 10

File: group-actors/src/main/java/org/sunbird/actors/CreateGroupActor.java [hunk 6]
```diff
@@ -24,7 +23,7 @@
 )
 public class CreateGroupActor extends BaseActor {
 
-  private GroupService groupService = GroupServiceImpl.getInstance();
+  private GroupService groupService = new GroupServiceImpl();
```
Comment (indrajra) [inline group-actors/src/main/java/org/sunbird/actors/CreateGroupActor.java:26]: Initing here will instantiate only for the number of actor instances - not for every API. So this is  buggy.

And secondly, we need to pass on the requestId MDC context - recall the ground-work did earlier. We must put to use here.

File: group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java (hunk 2 above)
Comment (aishwa8141) [reply group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java:49]: Yes, readGroup method returns group object, then I am converting into a string and saving into cache. But to put in the  response, it should be a MAp<String , Object> I cannot add groupObject so I am converting into Map at 70

File: group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java (hunk 3 above)
Comment (aishwa8141) [reply group-actors/src/main/java/org/sunbird/actors/ReadGroupActor.java]: done

File: service/conf/application.conf (hunk 4 above)
Comment (aishwa8141) [reply service/conf/application.conf]: redis.host=${sunbird_redis_host}
redis.port=${sunbird_redis_port}
redis.maxConnections=${?sunbird_redis_maxConnections}

File: service/conf/application.conf (hunk 5 above)
Comment (aishwa8141) [reply service/conf/application.conf]: done and added dispatcher in each Actor

File: group-actors/src/main/java/org/sunbird/actors/CacheActor.java (hunk 1 above)
Comment (aishwa8141) [reply group-actors/src/main/java/org/sunbird/actors/CacheActor.java:43]: When I call this Actor, I am setting reqId in the header of the request object


`[2020-07-16 17:50:07,269] [INFO] org.sunbird.actors.CacheActor:92 [] [thisActorSystem-akka.actor.default-dispatcher-5] [e8c689b4-10f9-4633-b321-6ffb1786a22b] - getCache started at 1594902007265 
[2020-07-16 17:50:07,271] [INFO] org.sunbird.actors.CacheActor:92 [] [thisActorSystem-akka.actor.default-dispatcher-5] [e8c689b4-10f9-4633-b321-6ffb1786a22b] - getCache method call 52d1ae62-42b0-4242-b271-43ba702b9f5a 
[2020-07-16 17:50:07,271] [INFO] org.sunbird.actors.CacheActor:92 [] [thisActorSystem-akka.actor.default-dispatcher-5] [e8c689b4-10f9-4633-b321-6ffb1786a22b] - getCache ended at 1594902007266 
`

**Pull request merged**